llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/InParallelToHAL.cpp

```diff
@@ -0,0 +1,141 @@
+
+  // TODO: Ideally only do this on buffers but we can't atm.
+  // Bufferize happens at the IREE level on HAL operations, we cannot just
+  // call the linalg_transform.bufferize operation here.
+  // Instead it happens automatically at the end of the linalg-transform-interp
+  // pass.
+
+  // TODO: #of enclosing InParallelOp determine the #idx in:
+  //   hal.interface.workgroup.id[#idx] : index
+  //   hal.interface.workgroup.count[#idx] : index
+  unsigned numEnclosingInParallelOps = 0;
+
+  // If inParallelOp.num_threads() is already a HAL op, stop applying.
+  Operation *numThreadOp = inParallelOp.num_threads().getDefiningOp();
+  if (numThreadOp &&
+      numThreadOp->getName().getStringRef() == kFakeHALWorkgroupIdOpName)
+    return failure();
+
+  // Rewriter-based RAUW operates on Operation* atm, bail if we can't get it.
+  Operation *numThreadDefiningOp = inParallelOp.num_threads().getDefiningOp();
+  if (!numThreadDefiningOp)
+    return failure();
+
+  Location loc = inParallelOp.getLoc();
+
+  // Custom hal.executable.entry_point.
+  auto region = std::make_unique<Region>();
+  Block &block = region->emplaceBlock();
+  {
+    // TODO: getOrCreate at top-level when multiple InParallelOp are used and
+    // replace the corresponding return.
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointToStart(&block);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Operation *op = rewriter.clone(*numThreadOp);
+    OperationState entryPointReturnOpState(
+        loc, kFakeHALReturnOpName, ValueRange{op->getResult(0), one, one},
+        TypeRange{}, ArrayRef<NamedAttribute>{}, BlockRange{}, {});
+    rewriter.createOperation(entryPointReturnOpState);
+  }
+  OperationState entryPointOpState(
+      loc, kFakeHALEntryPointRegionOpName, ValueRange{}, TypeRange{},
+      ArrayRef<NamedAttribute>{}, BlockRange{}, {region});
+  Operation *entryPointOp = rewriter.createOperation(entryPointOpState);
+
+  // Custom hal.workgroup.id.
+  OperationState idOpState(
+      loc, kFakeHALWorkgroupIdOpName, ValueRange{},
+      TypeRange{rewriter.getIndexType()},
+      ArrayRef<NamedAttribute>{
+          NamedAttribute(rewriter.getStringAttr("idx"),
+                         rewriter.getIndexAttr(numEnclosingInParallelOps))});
+  Operation *idOp = rewriter.createOperation(idOpState);
+
+  // Custom hal.workgroup.count.
+  OperationState countOpState(
+      loc, kFakeHALWorkgroupCountOpName, ValueRange{},
+      TypeRange{rewriter.getIndexType()},
+      ArrayRef<NamedAttribute>{
+          NamedAttribute(rewriter.getStringAttr("idx"),
+                         rewriter.getIndexAttr(numEnclosingInParallelOps))});
+  Operation *countOp = rewriter.createOperation(countOpState);
+
+  // Get a reference to the terminator that will subsequently be moved.
+  PerformConcurrentlyOp performConcurrentlyOp = inParallelOp.getTerminator();
+
+  // First, update the uses of num_threads() within the inParallelOp block.
```

Comment:
Is `num_threads()` a function call? Why not use an op that is `linalg_ext.num_threads`, then the transformation of that to `hal.workgroup.count` is straight-forward.
Dont you need a similar `thread_id()` as well?

Reply:
That is the name of the operand in the InParallelOp, this is just the accessor.